src/DataValues/MatrixValue.php

```diff
@@ -0,0 +1,128 @@
+<?php
+namespace DataValues;
+
+use DataValues\DecimalValue;
+/**
+ * Class representing a matrix.
+ *
+ * @licence GNU GPL v2+
+ * @author Andrius Merkys < andrius.merkys@gmail.com >
+ */
+class MatrixValue extends DataValueObject {
+
+	/**
+	 * @var array[]
+	 */
+	private $value;
+
+	/**
+	 * Constructs a new MatrixValue object, representing the given value.
+	 *
+	 * @var array[] $value 2-dimensional array of numbers (int|float|string|DecimalValue).
+	 *
+	 * @throws IllegalValueException
+	 */
+	public function __construct( $value ) {
+		if ( !is_array( $value ) ) {
+			throw new IllegalValueException( '$value must be an array' );
+		}
+
+		if ( count( $value ) == 0 ) {
+			throw new IllegalValueException( '$value can not be an empty array' );
+		}
+
+		$value_now = array();
+
+		$columns = null;
+		foreach ( $value as $row ) {
+			if( !is_array( $row ) ) {
+				throw new IllegalValueException( '$value must be a 2-dimensional array' );
+			}
+			if( $columns == null ) {
+				$columns = count( $row );
+			} else if( $columns != count( $row ) ) {
+				throw new IllegalValueException( 'all rows of $value must be of the same length' );
+			}
+
+			/**
+			 * Filtering the matrix, converting int|float|string into
+			 * DecimalValue.
+			 */
```

Comment:
As a general rule, try to avoid nesting control structures or having multiple sequentially when they each need/introduce different local state. It often makes sense to do one of these regardless, however if you violate that guideline several times in a single method, it is very likely you went too far. Try to have all code directly in a method be on the same level of abstraction, one below that of the method itself.